cpp/src/arrow/compute/kernels/util-internal.cc

```diff
@@ -165,46 +171,96 @@ Datum WrapDatumsLike(const Datum& value, const std::vector<Datum>& datums) {
-                                            Datum* out) {
-  std::vector<std::shared_ptr<Buffer>> data_buffers;
-  const ArrayData& in_data = *input.array();
-  MemoryPool* pool = ctx->memory_pool();
+Status PropagateNulls(FunctionContext* ctx, const ArrayData& input, ArrayData* output) {
+  const int64_t length = input.length;
 
-  // Handle the validity buffer.
-  if (in_data.offset == 0 || in_data.null_count <= 0) {
-    // Validity bitmap will be zero copied (or allocated when buffer is known).
-    data_buffers.emplace_back();
-  } else {
+  if (output->buffers.size() == 0) {
+    // Ensure we can assign a buffer
+    output->buffers.resize(1);
+  }
+
+  // Handle validity bitmap
+  output->null_count = input.GetNullCount();
+  if (input.offset != 0 && output->null_count > 0) {
+    DCHECK(input.buffers[0]);
+    const Buffer& validity_bitmap = *input.buffers[0];
     std::shared_ptr<Buffer> buffer;
-    RETURN_NOT_OK(AllocateBitmap(pool, in_data.length, &buffer));
+    RETURN_NOT_OK(ctx->Allocate(BitUtil::BytesForBits(length), &buffer));
     // Per spec all trailing bits should indicate nullness, since
     // the last byte might only be partially set, we ensure the
     // remaining bit is set.
     ZeroLastByte(buffer.get());
     buffer->ZeroPadding();
-    data_buffers.push_back(buffer);
+    internal::CopyBitmap(validity_bitmap.data(), input.offset, length,
+                         buffer->mutable_data(), 0 /* destination offset */);
+    output->buffers[0] = std::move(buffer);
+  } else {
+    output->buffers[0] = input.buffers[0];
   }
-  // Allocate the boolean value buffer.
+  return Status::OK();
+}
+
+Status PrimitiveAllocatingUnaryKernel::Call(FunctionContext* ctx, const Datum& input,
+                                            Datum* out) {
+  std::vector<std::shared_ptr<Buffer>> data_buffers;
+  const ArrayData& in_data = *input.array();
+
+  DCHECK_EQ(out->kind(), Datum::ARRAY);
+
+  ArrayData* result = out->array().get();
+
+  result->buffers.resize(2);
+
+  const int64_t length = in_data.length;
+
+  // Allocate the value buffer
   std::shared_ptr<Buffer> buffer;
-  RETURN_NOT_OK(AllocateBitmap(pool, in_data.length, &buffer));
-  // Some utility methods access the last byte before it might be
-  // initialized this makes valgrind/asan unhappy, so we proactively
-  // zero it.
-  ZeroLastByte(buffer.get());
-  data_buffers.push_back(buffer);
-  out->value = ArrayData::Make(null(), in_data.length, data_buffers);
+  if (out_type_->id() != Type::NA) {
+    const auto& fw_type = checked_cast<const FixedWidthType&>(*out_type_);
+
+    int bit_width = fw_type.bit_width();
+    int64_t buffer_size = 0;
 
+    if (bit_width == 1) {
+      buffer_size = BitUtil::BytesForBits(length);
+    } else if (bit_width % 8 == 0) {
+      buffer_size = length * fw_type.bit_width() / 8;
+    } else {
+      DCHECK(false);
+    }
+    RETURN_NOT_OK(ctx->Allocate(buffer_size, &buffer));
+    buffer->ZeroPadding();
+
+    if (bit_width == 1) {
```

Comment:
i left a bug in here that ASAN caught, we should also check that buffer_size > 0 (I can file a JIRA and submit a PR separately if you want me to handle it).

Reply:
done. Can you add a unit test as a follow up?